### PR TITLE
Creating file structure on sites

### PR DESCRIPTION
**Pull request by zlatkoHrastic**
- Added the ability to scan all the sites and fill the default document library on a site with data
- Site is filled with about 1000 items
- Site is only filled with items if the default docLib contains  less than 500 items (this allows us the "continue on fail" feature)
- Direct permissions and sharing links are generated
- Uploaded files can be opened with their respective tools, but they are empty

### CODE REVIEW

**Review by matijahanzic — APPROVED** [review]

File: SysKit.ODG.App/SysKit.ODG.Office365Service/GraphApiManagers/GroupGraphApiClient.cs [hunk 1]
```diff
@@ -265,6 +265,22 @@ public async Task<List<string>> GetAllTenantTeamIds()
             return membersLookup;
         }
 
+        /// <inheritdoc />
+        public async Task<List<string>> GetAllTenantGroupEmails()
+        {
+            var groupEmailRequest = _graphServiceBetaClient.Groups.Request().Select(g => g.Mail).Top(999);
```
Comment (tomislavkunaj) [inline SysKit.ODG.App/SysKit.ODG.Office365Service/GraphApiManagers/GroupGraphApiClient.cs:271]: Možda samo naznači nekako da ne dohvaćaš sve mejlove nego samo 999 njih? Čisto da nekome u budućnosti ne bude neočekivano ponašanje.

**Review by tomislavkunaj — APPROVED** [review]

File: SysKit.ODG.App/SysKit.ODG.Office365Service/GraphApiManagers/GroupGraphApiClient.cs (hunk 1 above)
Comment (zlatkoHrastic) [reply SysKit.ODG.App/SysKit.ODG.Office365Service/GraphApiManagers/GroupGraphApiClient.cs:271]: It still fetches everything. Top is a bad name for this parameter, it is basically page size.
https://docs.microsoft.com/en-us/graph/query-parameters#top-parameter

File: SysKit.ODG.App/SysKit.ODG.DataGeneration/Groups/GroupDataGeneration.cs [hunk 2]
```diff
@@ -179,5 +184,188 @@ private void populateSampleGroupProperties(GroupEntry groupEntry, IUserEntryColl
...
+
+            var sharingLinks = new List<SharingLinkEntry>();
+            if (hasALink)
+            {
+                sharingLinks.Add(getSharingLink(userEntryCollection));
+            }
+
+            if (hasASecondLink)
+            {
+                sharingLinks.Add(getSharingLink(userEntryCollection));
```
[the remark this answers is not in the document]
Comment (zlatkoHrastic) [reply SysKit.ODG.App/SysKit.ODG.DataGeneration/Groups/GroupDataGeneration.cs:327]: Error handling is done per item, and Microsoft should merge same link types so this is not necesarry

File: SysKit.ODG.App/SysKit.ODG.DataGeneration/Groups/GroupDataGeneration.cs [hunk 3]
```diff
@@ -179,5 +184,188 @@ private void populateSampleGroupProperties(GroupEntry groupEntry, IUserEntryColl
...
             groupEntry.Template = memberAndOwnerGenerationResult.Template;
         }
+
+        public List<ContentEntry> GenerateDocumentsFolderStructure(int itemsPerSite, IUserEntryCollection userEntryCollection, List<string> groupEmails)
+        {
+            var itemCounter = 0;
+            var folderCounter = 0;
+            var rootNodes = new List<ContentEntry>();
+
+            var numberOfRootFolders = RandomThreadSafeGenerator.Next(4, 6);
```
[the remark this answers is not in the document]
Comment (zlatkoHrastic) [reply SysKit.ODG.App/SysKit.ODG.DataGeneration/Groups/GroupDataGeneration.cs:194]: I can do naming with the available time

**Pull request merged**